File: parametric_ai.cpp
#include <algorithm>
#include <cassert>
#include <cmath>
#include <functional>
#include <iostream>
#include <limits>
#include <numeric>
#include <queue>
#include <type_traits>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;

template <typename R, typename V>
using Graph = unordered_map<V, unordered_map<V, R>>;

template <typename R> class NegCycleFinder {
public:
  NegCycleFinder(const Graph<R, int> &graph) : graph_(graph) {}

  vector<int> find_neg_cycle(unordered_map<int, R> &dist,
                             function<R(pair<int, int>)> get_weight) {
    int n = graph_.size();
    vector<int> prev(n, -1);
    vector<int> cycle;
    int last = -1;
    for (int i = 0; i < n; ++i) {
      for (auto &[u, edges] : graph_) {
        for (auto &[v, _] : edges) {
          if (dist[u] + get_weight({u, v}) < dist[v]) {
            dist[v] = dist[u] + get_weight({u, v});
            prev[v] = u;
            last = v;
          }
        }
      }
    }
    if (last != -1) {
      for (int i = 0; i < n; ++i) {
        last = prev[last];
      }
      for (int v = last;; v = prev[v]) {
        cycle.push_back(v);
        if (v == last && cycle.size() > 1) {
          break;
        }
      }
      reverse(cycle.begin(), cycle.end());
    }
    return cycle;
  }

private:
  const Graph<R, int> &graph_;
};

/**
 * @brief
 *
 * @tparam R
 */
template <typename R> class Omega {
public:
  /**
   * @brief
   *
   * @param ratio
   * @param edge
   * @return R
   */
  R distance(R ratio, pair<int, int> edge) const {
    return log(ratio * exp(weights_[edge]) +
               (1 - ratio) * exp(-weights_[edge]));
  }

  /**
   * @brief
   *
   * @param cycle
   * @return R
   */
  R zero_cancel(const vector<int> &cycle) const {
    R sum = 0;
    for (int i = 0; i < cycle.size(); ++i) {
      sum += weights_.at({cycle[i], cycle[(i + 1) % cycle.size()]});
    }
    return exp(sum);
  }

  /**
   * @brief Set the weight object
   *
   * @param edge
   * @param weight
   */
  void set_weight(pair<int, int> edge, R weight) { weights_[edge] = weight; }

private:
  unordered_map<pair<int, int>, R> weights_; ///< weights
};

template <typename R, typename V>
using MutableGraph = unordered_map<V, unordered_map<V, R>>;

template <typename R, typename V>
using Mapping = unordered_map<V, unordered_map<V, R>>;

template <typename R, typename V>
using MutableMapping = unordered_map<V, unordered_map<V, R>>;

template <typename R, typename V> using Cycle = vector<pair<V, V>>;

/**
 * @brief
 *
 * @tparam R
 * @tparam V
 * @param gra
 * @param ratio
 * @param omega
 * @param dist
 * @return R
 */
template <typename R, typename V>
R max_parametric(const Mapping<R, V> &gra, R ratio, const Omega<R> &omega,
                 MutableMapping<R, V> &dist) {
  NegCycleFinder<R> ncf(gra);
  R r_min = ratio;
  Cycle<R, V> c_min;
  Cycle<R, V> cycle;
  while (true) {
    for (auto &ci : ncf.find_neg_cycle(dist, [&](pair<V, V> edge) {
           return omega.distance(ratio, edge);
         })) {
      R ri = omega.zero_cancel(ci);
      if (r_min > ri) {
        r_min = ri;
        c_min = ci;
      }
    }
    if (r_min >= ratio) {
      break;
    }
    cycle = c_min;
    ratio = r_min;
  }
  return ratio;
}

File: parametric_ai.cs
using System.Collections.Generic;
using System.Linq;

public interface ParametricAPI<V, R> {
  R Distance(R ratio, (V, V)e);
  R ZeroCancel(List<(V, V)> cycle);
}

public class MaxParametricSolver<V, R> {
  private readonly Dictionary<V, Dictionary<V, dynamic>> gra;
  private readonly ParametricAPI<V, R> omega;
  private readonly NegCycleFinder<V, R> ncf;

  public MaxParametricSolver(Dictionary<V, Dictionary<V, dynamic>> gra,
                             ParametricAPI<V, R> omega) {
    this.gra = gra;
    this.omega = omega;
    this.ncf = new NegCycleFinder<V, R>(gra);
  }

  public (R, List<(V, V)>) Run(Dictionary<V, R> dist, R ratio) {
    R r_min = ratio;
    List<(V, V)> c_min = new List<(V, V)>();
    List<(V, V)> cycle = new List<(V, V)>();
    while (true) {
      foreach (var ci in ncf.FindNegCycle(dist,
                                          e => omega.Distance(ratio, e))) {
        R ri = omega.ZeroCancel(ci);
        if (r_min.CompareTo(ri) > 0) {
          r_min = ri;
          c_min = ci;
        }
      }
      if (r_min.CompareTo(ratio) >= 0) {
        break;
      }
      cycle = c_min;
      ratio = r_min;
    }
    return (ratio, cycle);
  }
}

public class NegCycleFinder<V, R> {
  private readonly Dictionary<V, Dictionary<V, dynamic>> gra;

  public NegCycleFinder(Dictionary<V, Dictionary<V, dynamic>> gra) {
    this.gra = gra;
  }

  public List<List<(V, V)>> FindAllNegCycles() {
    var dist = new Dictionary<V, R>();
    var prev = new Dictionary<V, (V, V)>();
    foreach (var u in gra.Keys) {
      dist[u] = default(R);
      prev[u] = (default(V), default(V));
    }
    var cycle = new List<(V, V)>();
    foreach (var u in gra.Keys) {
      if (dist[u].CompareTo(default(R)) != 0) {
        continue;
      }
      var v = u;
      while (true) {
        var e = (default(V), default(V));
        foreach (var (to, w) in gra[v]) {
          if (dist[to].CompareTo(dist[v].Add(w)) > 0) {
            dist[to] = dist[v].Add(w);
            prev[to] = (v, to);
            e = (v, to);
          }
        }
        if (e.Item1 == null) {
          break;
        }
        v = e.Item2;
        if (cycle.Contains(e)) {
          var idx = cycle.IndexOf(e);
          return new List<List<(V, V)>> { cycle.Skip(idx).ToList() };
        }
        cycle.Add(e);
      }
      foreach (var e in cycle) {
        if (dist[e.Item2].CompareTo(default(R)) != 0) {
          continue;
        }
        var to = e.Item2;
        while (true) {
          to = prev[to].Item1;
          if (dist[to].CompareTo(default(R)) != 0) {
            break;
          }
          dist[to] = default(R);
        }
      }
      cycle.Clear();
    }
    return new List<List<(V, V)>>();
  }

  public List<List<(V, V)>> FindNegCycle(Dictionary<V, R> dist,
                                         System.Func<(V, V), R> weight) {
    var gra = new Dictionary<V, Dictionary<V, dynamic>>();
    foreach (var (u, v) in dist.Keys) {
      if (!gra.ContainsKey(u)) {
        gra[u] = new Dictionary<V, dynamic>();
      }
      gra[u][v] = weight((u, v));
    }
    foreach (var (u, to) in gra.Keys.SelectMany(
                 u => gra[u].Keys.Select(to => (u, to)))) {
      gra[u][to] = gra[u][to].Subtract(dist[u]).Add(dist[to]);
    }
    return new NegCycleFinder<V, R>(gra).FindAllNegCycles();
  }
}
